### DISCUSSION TITLE
Why does storage::Client::ListObjects not support pageToken?

### DISCUSSION THREAD

**Discussion by mchelnokov** [googleapis/google-cloud-cpp#14060] [General]
Hi,
The GCS has [a method](https://cloud.google.com/storage/docs/json_api/v1/objects/list) to retrieve list of objects in a bucket. This method has pageToken parameter, to paginate through the results using a token. The token [is supported](https://github.com/googleapis/google-cloud-cpp/blob/8fe023c04a8bea09bd0b21506277955000a0df74/google/cloud/storage/internal/object_requests.h#L79) by storage::internal::ListObjectsResponse, but then it [isn't used](https://github.com/googleapis/google-cloud-cpp/blob/8fe023c04a8bea09bd0b21506277955000a0df74/google/cloud/storage/client.h#L1043) by storage::client::ListObjects. The startOffset parameter isn't the same as the pageToken, it is to filter results lexicographically, not for pagination.

**Comment by coryan** [comment]
Yes, that is correct.  We are missing a way to paginate (as opposed to iterate) over objects (and buckets).  However, `startOffset` is more useful than it may seem.  Objects are returned lexicographically by ListObjects (with or without pagination), so starting from the last object you saw gives you almost the same effect.

https://github.com/googleapis/google-cloud-cpp/issues/5742 was an attempt to fix this for buckets, but we did not get very far.
